### PR TITLE
Add typing for beakerlib.by

### PR DESCRIPTION
**Pull request by thrix**
Signed-off-by: Miroslav Vadkerti <mvadkert@redhat.com>

### CODE REVIEW

File: tmt/beakerlib.py [hunk 1]
```diff
@@ -236,7 +272,11 @@ def fetch(self):
                     f"for a deep library ({error}).")
 
 
-def dependencies(original_require, original_recommend=None, parent=None):
+def dependencies(
+    original_require: List[str],
+    original_recommend: Optional[List[str]] = None,
+    parent: Optional[CommonWithLibraryCache] = None
```
Comment (happz) [inline tmt/beakerlib.py:278]: I think this will lead to `CommonWithLibraryCache` leaking outside of this module. Whoever calls `dependencies()` function would have to apply `cast(CommonWithLibraryCache, ...)` to arguments passed to `dependencies()`.

I'd recommend keeping `CommonWithLibraryCache` fully inside the new `_library_cache` property, while `parent` would remain `Common`, compatible with the rest of the codebase. That way, the fact there is a special attribute involved would be hidden from anything outside of `Library._library_cache` property.

File: tmt/beakerlib.py [hunk 2]
```diff
@@ -125,19 +146,26 @@ def __init__(self, identifier, parent=None):
...
+        if not hasattr(self.parent, '_library_cache'):
+            self.parent._library_cache = dict()
+
+        return self.parent._library_cache
+
+    def fetch(self) -> None:
         """ Fetch the library (unless already fetched) """
         # Initialize library cache (indexed by the repository name)
         if not hasattr(self.parent, '_library_cache'):
             self.parent._library_cache = dict()
```
Comment (happz) [inline tmt/beakerlib.py:164]: These two lines are now part of `_library_cache` property, therefore can be dropped.

File: tmt/beakerlib.py (hunk 1 above)
Comment (thrix) [reply tmt/beakerlib.py:278]: ok

File: tmt/beakerlib.py (hunk 2 above)
Comment (thrix) [reply tmt/beakerlib.py:164]: ok

File: tmt/beakerlib.py [hunk 3]
```diff
@@ -100,20 +120,21 @@ def __init__(self, identifier, parent=None):
...
-            if not self.repo and self.path:
+                repo = repo_search.group(1)
+            if not repo and self.path:
                 try:
-                    self.repo = os.path.basename(self.path)
+                    repo = os.path.basename(self.path)
                 except TypeError:
                     raise tmt.utils.GeneralError(
                         f"Unable to parse repository name from '{self.path}'.")
+            assert repo
```
Comment (happz) [inline tmt/beakerlib.py:136]: This is suspicious. At least one of `self.url` and `self.path` shall be set, according to line 106. Ok, let's say it's `self.url` and `self.path` is `None`. Then line 123, `repo` is read from `identifier`, let's say it's empty or `None`-ish - apparently, that can happen, because on line 124 there is a check for `not repo`. `self.url` is set, so we dive under the `if`, and let's say `repo_search` is valid, not `None`, but `repo_search.group(1)` is an empty string - can this happen? If so, then `not repo and self.path` condition on line 130 will not trigger, because `self.path` is `None` as we already know, and `repo` will not pass the `assert repo` check you added.

In other words, this `assert` suggests the code flow has some shady area which prevents mypy from deducing `repo` is non-empty value by now. If `repo == ""` is allowed (can it happen on line 129?), then this assert shall be `assert repo is not None`. Or, if the block on like 124 fails to produce non-empty `repo`, with `self.path is None`, the check on like 130 needs an update.

File: tmt/beakerlib.py [hunk 4]
```diff
@@ -167,6 +193,7 @@ def fetch(self):
                         ['git', 'clone', self.url, directory],
                         env={"GIT_ASKPASS": "echo"})
                 else:
+                    assert self.path
```
Comment (happz) [inline tmt/beakerlib.py:196]: I'd propose `assert self.url or self.path` before line 191, to signal that at least one of them shall be set, and it should produce consistent.

Comment (thrix) [reply tmt/beakerlib.py:196]: so add another assert @happz ? I need to keep this here otherwise it is not happy with just that at all

Comment (thrix) [reply tmt/beakerlib.py:196]: added rather if / else

File: tmt/beakerlib.py (hunk 3 above)
Comment (happz) [reply tmt/beakerlib.py]: I see. Then maybe the condition on line 130 should be `elif`: if `repo` is either already set or comes set from the block above, then `if not repo` can never trigger. Or maybe even refactored, to help mypy (and, obviously, me :), to something like:

```python
if not repo:
  if self.url:
    ...
  else:
   ...

# assert repo - maybe this one won't be needed
self.repo = repo
```

Right, so with this it makes sense, since mypy cannot know the regex won't let this happen. With the refactoring, maybe the assert won't be needed at all, but it's been certainly a bit complicated to me.

**Comment by thrix** [comment]
Cool, all tests pass :rocket:

**Review by happz — APPROVED** [review]

**Comment by happz** [comment]
LGTM!

**Review by ukulekek — APPROVED** [review]

File: tmt/beakerlib.py [hunk 5]
```diff
@@ -100,20 +120,24 @@ def __init__(self, identifier, parent=None):
...
+            elif not repo and self.path:
                 try:
-                    self.repo = os.path.basename(self.path)
+                    repo = os.path.basename(self.path)
                 except TypeError:
                     raise tmt.utils.GeneralError(
                         f"Unable to parse repository name from '{self.path}'.")
+            else:
+                raise tmt.utils.GeneralError(
+                    f"Unable to parse repository name from '{self.path}'.")
```
Comment (psss) [inline tmt/beakerlib.py:138]: Why are you adding this extra `else` branch? This seems to break `/tests/discover/libraries/` as `repo` can also be detected from `nick` (alread set before the condition).

File: tmt/beakerlib.py [hunk 6]
```diff
@@ -154,23 +182,27 @@ def fetch(self):
                     f"using ref '{library.ref}'.")
             self.parent.debug(f"Library '{self}' already fetched.", level=3)
             # Reuse the existing metadata tree
-            self.tree = library.tree
+            self.tree: fmf.Tree = library.tree
         # Fetch the library and add it to the index
         except KeyError:
             self.parent.debug(f"Fetch library '{self}'.", level=3)
             # Prepare path, clone the repository, checkout ref
+            assert self.parent.workdir
```
Comment (psss) [inline tmt/beakerlib.py:190]: What's the point of this assert?

**Review by psss — CHANGES_REQUESTED** [review]
Thanks for working on the typing. Looks good in general. However, the slight change in the `repo` name detection breaks `nick`. The `/tests/discover/libraries/` test is disable in the CI as it needs full virtualization. We should probably enable at least subset of its tests which do not require vm in th CI. Some nitpick adjustments commited in 3f251df.

File: tmt/beakerlib.py (hunk 5 above)
Comment (thrix) [reply tmt/beakerlib.py:138]: hmm, I will need to check

File: tmt/beakerlib.py (hunk 6 above)
Comment (thrix) [reply tmt/beakerlib.py:189]: @psss  Because `self.parent.workdir` property defines the return type `Optional[str]`, this is required here to make `mypy` happy on line 190, where `join` does not take a None value. 

The other solution would be:
* to check it for `None` here and raise an exception
* fix the `workdir` property that it cannot return `None`

So?

File: tmt/beakerlib.py (hunk 5 above)
Comment (thrix) [reply tmt/beakerlib.py:138]: Indeed, thanks, updated. The thing is, `nick` can be `None`, so `mypy` complains if that can happen. I changed the check to check for that and return a (I hope) resonable error.

File: tmt/beakerlib.py (hunk 6 above)
Comment (happz) [reply tmt/beakerlib.py:189]: Can `self.parent.workdir` be `None` at this point, is it allowed, or does code flow absolutely rules such an option out? mypy cannot tell, `workdir` can be `None` and `None` cannot be passed to `join()`, but if humans *know* for sure `workdir` will no longer be none when tmt already works with libraries, then `assert` "narrows" the type for mypy.

An exception is a valid tool as well, changing `workdir` to never allow `None` might be complicated, I recall seeing some `workdir=None` parameters around in `tmt.utils.Common`.

Comment (thrix) [reply tmt/beakerlib.py:190]: yeah, I would leave it to the expert :) let's wait

Comment (psss) [reply tmt/beakerlib.py:190]: @thrix, @happz, thanks for the detailed clarification. So here, the `workdir` will always be defined as we get `parent` from a real `discover.fmf` plugin instance. Also, if the `parent` is not provided, `Common(workdir=True)` is used which also ensures that a workdir is initialized. So, I guess, the `assert` way seems to be the most natural approach here.

**Review by psss — APPROVED** [review]
The extra condition branches which were added were unreachable because either `url` or `path` has to be always set which is checked at the beginning. I've covered this by 68b1f4d. Please, have a look if everything's ok. And thanks to @happz for the help with investigating this!

**Comment by thrix** [comment]
@psss are we missing some tests then? The correct functionality really should be caught by tests.

**Comment by psss** [comment]
Well, we don't have coverage for each and every code path here, and I don't think it would make sense to invest substantial amounts of time to have 100% line coverage. Those conditions branches you introduced where just unreachable, no harm, but also no benefit. They made it hard to understand what's actually happening and what the logic is.

**Review by happz — APPROVED** [review]
